### DISCUSSION TITLE
Problems connecting to server

### DISCUSSION THREAD

**Discussion by alzeha** [UPPAALModelChecker/UPPAAL-Meta#109] [Q&A] [answered]
Hi,

my name is Alexander, I'm a PhD student at TU Darmstadt/Germany.

Am I the only one, having trouble setting up a remote verification server?

I want to use my desktop (running linux mint) as remote server. I run (via ssh):

```
# reboot
$ cd /path/to/uppaal64-4.1.26-1/bin-Linux/
$ ./socketserver
UPPAAL Socket Server
Copyright (c) 1995 - 2000, Aalborg University and Uppsala University
All rights reserved.
Listening on port 2350
```

I tried from both, archlinux and windows, to locally run uppaal and connect to the server. E.g. at windows I run:

`uppaal64-4.1.26-1> java -jar .\uppaal.jar --serverHost <loc_ip> --serverPort 2350`

An error window pops up:

```
Unknown protocol version: error
Java version: 11.0.15
Java vendor: Microsoft
JVM version: 11.0.15+10-LTS
JVM vendor: Microsoft
JVM name: OpenJDK 64-Bit Server VM
JRE version: 11
JRE vendor: Oracle Corporation
JRE name: Java Platform API Specification
OS name: Windows 10
OS arch: amd64
```

at server side:

```
Connection from <laptop> accepted
Connection from <laptop> closed
```

after clicking "ok", uppaal opens anyway. But starting a verification is not possible. A window opens, asking for a License installation (??? i never read about sth like this?). 

This happens also when I start the socketserver as sudo.

Is there anywhere a more clear tutorial, regarding the question what i have to do to start a remote server? I don't get whats happening here.

Sorry for the inconvenience.

Best regards,
Alexander

**Comment by RasmusRendal** [comment]
Hello Alexander. Could you try running the socketserver with the arguments `-dm`, as in `./socketserver -dm`, and then sending the output? This should output a server trace, which should be helpful in troubleshooting.

**Comment by alzeha** [comment]
Hi,

thanks for your answer!

The result is

Caught signal 17, closing proxy.
1697: ERROR /usr/bin/env: symbol lookup error: path/to/uppaal64-4.1.26-1/bin-Linux/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE

huh, so you're trying to statically link against a shipped libc, right? Is this really a good idea?

Something is breaking at this point... I'm unsure what exactly the problem is.

**Reply by mikucionisaau** [reply]
> Caught signal 17, closing proxy. 1697: ERROR /usr/bin/env: symbol lookup error: path/to/uppaal64-4.1.26-1/bin-Linux/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE
> 
> huh, so you're trying to statically link against a shipped libc, right? Is this really a good idea?

It's dynamic linking. I suspect that the `libc` on your system is newer, thus the shell wrapper fails to run `/usr/bin/env` with the `libc` from Uppaal.

Could you try running the socketserver and the server directly with your `libc`? i.e.:
```sh
./socketserver.bin -s ./server.bin
```
If the above works, you can just overwrite the shell wrappers with the binaries, like this:
```sh
mv server.bin server
mv socketserver.bin socketserver
```

**Reply by alzeha** [reply]
same problem:

$ ./socketserver.bin -s ./server.bin
./socketserver.bin: symbol lookup error: /path/to/uppaal64-4.1.26-1/bin-Linux/libc.so.6: undefined symbol: _dl_catch_error_ptr, version GLIBC_PRIVATE

**Reply by RasmusRendal** [reply]
Did you try socat as described further down?

**Comment by RasmusRendal** [comment, marked as answer]
So what is happening is that socketserver is running with the included linker and `LD_LIBRARY_PATH`, which we include to make sure that our distribution works on pretty much anything.

However, what the socketserver does is simply running `./server` in a loop and serving it over port 2350. But in this case `./server` is a shell script, which uses `/usr/bin/env` from your system, which isn't compatible with the included linker and `LD_LIBRARY_PATH`, resulting in the error. Fixing the socketserver might turn out to be a bit an ordeal, so until that is fixed, you can use socat instead of the socketserver as a workaround:

```
socat TCP-LISTEN:2350,reuseaddr,fork EXEC:"./server"
```

You might have to install it using apt: `sudo apt install socat` should do the trick.

**Reply by alzeha** [reply]
Thats working, thanks!

**Reply by mikucionisaau** [reply]
@alzeha I am glad this worked for you. 
Could you please tell us the `libc` version on your computer? It can be displayed by the following command:
```sh
dpkg -l libc6
```

**Reply by alzeha** [reply]
sure.

```
$ dpkg -l libc6

Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version         Architecture Description
+++-==============-===============-============-=================================
ii  libc6:amd64    2.31-0ubuntu9.9 amd64        GNU C Library: Shared libraries
ii  libc6:i386     2.31-0ubuntu9.9 i386         GNU C Library: Shared libraries
```
